### PR TITLE
arch: arc: let zephyr run in normal mode of arc

### PR DESCRIPTION
**Pull request by vonhust**
This PR is initial implementation of some kind of TEE support for ARC.

As the TEE design/framework in zephyr matures,  more work will be needed. 

It includes necessary commits to let zephyr run in normal mode of
arc processors equipped with SecureShield.

In this PR, zephyr can be built into:
1. imgage for secure mode (CONFIG_ARC_SECURE_FIRMWARE <- TRUSTED_EXECUTION_SECURE)
   * full access to system resource
   * full zephyr features are supported,  USERSPACE, stack protection
2. image for normal mode (CONFIG_ARC_NORMAL_FIRMWARE)
  * no mpu support in normal mode, i.e., no USERSPACE
  * to run this normal image,  it requires a secure bootloader or secure zephyr application to
     boot, as when processor is reset, it's secure mode.

To show how normal zephyr application works with secure zephyr applicaiton, an example is provided in samples/boards/arc_secure_services,  more details, pls go to the README.rst of this example.

### CODE REVIEW

File: arch/arc/core/fault_s.S
Comment (ruuddw) [inline arch/arc/core/fault_s.S]: I think it's actually ok to read ERSEC_STAT from Normal mode as well - fields without access will be 0 (IOW/RAZ)

File: arch/arc/core/prep_c.c
Comment (ruuddw) [inline arch/arc/core/prep_c.c]: Why? Normal mode has it's own vector table, shouldn't that be used (and updated) in Normal mode?

File: arch/arc/core/swap.S
Comment (ruuddw) [inline arch/arc/core/swap.S]: See earlier remark on ERSEC_STAT, I think it's ok to access from Normal mode as well.

File: arch/arc/core/thread.c
Comment (ruuddw) [inline arch/arc/core/thread.c]: Is reading the current contents best way to initialize the SEC_STAT in the initial context? Can't it be some constant value?

File: arch/arc/core/vector_table.c
Comment (ruuddw) [inline arch/arc/core/vector_table.c]: Assume this is for the temporary secure initialization code, it's not really 'logical' to declare secure stuff in a 'NORMAL_FIRMWARE' conditional section. (maybe add a 'TODO' note)

File: arch/arc/include/swap_macros.h
Comment (ruuddw) [inline arch/arc/include/swap_macros.h]: not sure we cover all combination of options, looks like HAS_SECURE implies we also have USERSPACE - no support for Secure but no userspace?

File: arch/arc/include/swap_macros.h
Comment (ruuddw) [inline arch/arc/include/swap_macros.h]: Missing case for HAS_SECURE and NORMAL_FIRMWARE?

File: arch/arc/include/v2/irq.h
Comment (ruuddw) [inline arch/arc/include/v2/irq.h]: not sure an sjli call would be required here, this setup is only done once and could be done by Secure firmware before switching to Normal Zephyr.

File: boards/arc/nsim_em/nsim_sem_normal.yaml
Comment (ruuddw) [inline boards/arc/nsim_em/nsim_sem_normal.yaml]: copy/paste typo: this should be nsim_sem_normal

File: drivers/interrupt_controller/arcv2_irq_unit.c
Comment (ruuddw) [inline drivers/interrupt_controller/arcv2_irq_unit.c]: maybe suspend/resume functionality should have different overall implementation, Secure firmware should probably also support suspend/resume for this to work.

File: include/arch/arc/v2/arcv2_irq_unit.h
Comment (ruuddw) [inline include/arch/arc/v2/arcv2_irq_unit.h]: All interrupts either Secure or Normal is probably too strong, some should be allocated to Secure and some to Normal depending on application. For now ok, but add a note (or issue) to support configuration/allocation of interrupts to Secure or Normal.

**Comment by nashif** [comment]
@vonhust @ruuddw can you please rebase and drive to closure? thanks.

**Comment by nashif** [comment]
@ruuddw are you ok with this?

**Comment by nashif** [comment]
> @nashif still work in progress, I've marked it DNM for now.

ok

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: Ping me when there's documentation to review...

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: arch/arc/Kconfig
Comment (ruuddw) [inline arch/arc/Kconfig]: "normal state" -> normal mode

File: arch/arc/core/fault_s.S
Comment (ruuddw) [inline arch/arc/core/fault_s.S]: Since this requires a sjli call in Normal mode, maybe better now to not use re-use the common handler?
Guess this code is not in a timing critical path, but avoiding a slji would be good. It's a powerfull/dangerous sjli call so should not be use-able by other normal code (implementation of the sjli needs to check the caller)

File: arch/arc/core/irq_manage.c
Comment (ruuddw) [inline arch/arc/core/irq_manage.c]: is this split required? This enforces secure priorities are higher than normal priorities. And, it requires hardware configuration with multiple priority levels (guess most ARC configs would have that anyway).

File: arch/arc/core/regular_irq.S
Comment (ruuddw) [inline arch/arc/core/regular_irq.S]: not all interrupts are secure, or return to secure mode - why set IRM?

File: arch/arc/core/secureshield/arc_secure.S
Comment (ruuddw) [inline arch/arc/core/secureshield/arc_secure.S]: preferably normal interrupts are allowed during secure calls.

File: arch/arc/core/secureshield/arc_secure.S
Comment (ruuddw) [inline arch/arc/core/secureshield/arc_secure.S]: Why is a SW dispatch table used instead of using multiple sjli calls and hw dispatching?

File: arch/arc/core/secureshield/arc_secure.S
Comment (ruuddw) [inline arch/arc/core/secureshield/arc_secure.S]: not just call id, also other arguments and - depending on the call id/secure system call feature requested - maybe even whether the caller is allowed to call the function.

File: arch/arc/core/secureshield/secure_sys_services.c
Comment (ruuddw) [inline arch/arc/core/secureshield/secure_sys_services.c]: I'd expect that Secure mode at system initialization defines certain interrupts as Normal, and rest as Secure. Not sure Normal mode needs this call.

File: arch/arc/core/swap.S
Comment (ruuddw) [inline arch/arc/core/swap.S]: is it actually necessary to set IRQ_ACT before doing a 'fake' rtie? Would be good if it can be avoided.

File: boards/arc/em_starterkit/arc_mpu_regions.c
Comment (ruuddw) [inline boards/arc/em_starterkit/arc_mpu_regions.c]: ultimately, peripherals should be split over Normal and Secure mode. Maybe normal and secure board dts files can be used for that.

File: samples/boards/arc_secure_services/src/main.c
Comment (ruuddw) [inline samples/boards/arc_secure_services/src/main.c]: is a thread required in Secure mode? We don't support mixed secure/normal threads (yet...), the Secure part is more like a library and initial entry point before handing over to Normal mode running full Zephyr.

File: arch/arc/Kconfig
Comment (ruuddw) [inline arch/arc/Kconfig]: interesting typo, can jump to Linux with a sjli call ;-)

Comment (vonhust) [reply arch/arc/Kconfig]: fixed

File: arch/arc/Kconfig
Comment (vonhust) [reply arch/arc/Kconfig]: fixed

File: arch/arc/core/fault_s.S
Comment (vonhust) [reply arch/arc/core/fault_s.S]: read is ok, but no write. It will make codes simpler

File: arch/arc/core/irq_manage.c
Comment (vonhust) [reply arch/arc/core/irq_manage.c]: I think it's required, or the normal mode can attack the secure mode through interrupt. With this assumption, the whole system is more secure and simpler, e.g. the cases of bounce of interrupts between secure and normal.  It's just something like fast irq's priority is higher then regular irq

File: arch/arc/core/regular_irq.S
Comment (vonhust) [reply arch/arc/core/regular_irq.S]: here, rtie is used to do a dispatch.  The interruptted context could be in normal mode and the target thread must be in secure mode. So the  IRM must be set to the correct value.

File: arch/arc/core/secureshield/arc_secure.S
Comment (vonhust) [reply arch/arc/core/secureshield/arc_secure.S]: it should be left to the implementation of secure system call handler.

File: arch/arc/core/reset.S
Comment (vonhust) [reply arch/arc/core/reset.S]: done

File: arch/arc/core/secureshield/arc_secure.S
Comment (vonhust) [reply arch/arc/core/secureshield/arc_secure.S]: normal mode can attack the secure mode through interrupt. So for more secure system, do not let normal world preempt the secure world

File: arch/arc/core/secureshield/secure_sys_services.c
Comment (vonhust) [reply arch/arc/core/secureshield/secure_sys_services.c]: There may be some dynamic cases.  If the interrupt allocation is static, yes, no need such a call.

File: arch/arc/core/swap.S
Comment (vonhust) [reply arch/arc/core/swap.S]: currently, no better way except we don't use rtie instruction. If not sec IRQ_ACT,  the rtie may call the exception return flow

File: arch/arc/core/thread.c
Comment (vonhust) [reply arch/arc/core/thread.c]: It could be  a constan value.  I will consider to change it later.

File: arch/arc/include/swap_macros.h
Comment (vonhust) [reply arch/arc/include/swap_macros.h]: if no userspace, no need to save and restore  USER/KERNEL_SP realted aux regs.

File: arch/arc/include/swap_macros.h
Comment (vonhust) [reply arch/arc/include/swap_macros.h]: fixed

File: boards/arc/em_starterkit/arc_mpu_regions.c
Comment (vonhust) [reply boards/arc/em_starterkit/arc_mpu_regions.c]: It depends on the application. Currently, I just let the peripherals shared between secure and normal mode

File: include/arch/arc/v2/arcv2_irq_unit.h
Comment (vonhust) [reply include/arch/arc/v2/arcv2_irq_unit.h]: It requires a mechinism to tell which interrupts are secure, which are normal

File: samples/boards/arc_secure_services/src/main.c
Comment (vonhust) [reply samples/boards/arc_secure_services/src/main.c]: It's a complicate case. It thread is not required, just disable the multithreading in Zephry. If we can let two zephyr run together, it will be simple for baremetal case.

File: arch/arc/core/irq_manage.c
Comment (ruuddw) [reply arch/arc/core/irq_manage.c]: Interrupt priorities are typically assigned based on real-time response requirements. That can be different from security concerns. Fast vs regular IRQ's is about response times, so different.

File: arch/arc/core/regular_irq.S
Comment (ruuddw) [reply arch/arc/core/regular_irq.S]: Why must target thread be in secure mode?

**Comment by nashif** [comment]
@ruuddw  @vonhust  still relevant, if yes, please update and rebase?

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: pls review the documentation.

**Comment by vonhust** [comment]
@nashif @ruuddw  I just rebase ARC TEE support on zephyr latest code. We can start the review again.
But i met a small problem reported by CI,  Dependcy loop issuse in kconfig. This issue also made "make htmldocs". I already took a look at this issue, didn't find a quick fix. Any ideas or hints?

**Comment by vonhust** [comment]
@ruuddw @abrodkin  I refactor this PR for 2.0  release. Although still very primitive,  it's a good start point for ARC TEE support

**Comment by ioannisg** [comment]
@vonhust is this PR abandoned?
If not, shall we move it to 2.1 release?

**Comment by vonhust** [comment]
> @vonhust is this PR abandoned?
> If not, shall we move it to 2.1 release?

It not abandoned,  i am trying to make it 2.0 in the left 2-3 days

**Comment by ioannisg** [comment]
> 
> 
> > @vonhust is this PR abandoned?
> > If not, shall we move it to 2.1 release?
> 
> It not abandoned, i am trying to make it 2.0 in the left 2-3 days

Oh cool! @vonhust So sorry,
I saw the conflicts and the PR "age" but didn't see you actually are working on it.

**Review by ioannisg — COMMENTED** [review]
I've only checked the generic arch code - that looks ok to me

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: Change to:
```
In the normal application, the MPU is not accessible, so there is
no user space and MPU-based stack checking.
```

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: Change to:
```
We'll use the  :ref:'dining-philosophers-sample' as an example
for the normal application.
```

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: Change this to:
```
Run using the bootloader
   The bootloader should load the secure and normal application into the correct place,
   then jump to the entry of the secure application.

Run using the debugger (recommended)
   Use the gdb debugger to load and run the two applications.
```

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: delete ``please``

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: Change to:
```
For nsim sem, you need two consoles: one for
application output, and one for the debugger.
```

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: ``for the debugger``

File: samples/boards/arc_secure_services/README.rst
Comment (dbkinder) [inline samples/boards/arc_secure_services/README.rst]: do you need to specify the path of the secure application, or is that on line 120?

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: arch/arc/core/reset.S
Comment (vonhust) [reply arch/arc/core/reset.S]: It's removed in latest codes

**Review by ruuddw — APPROVED** [review]
There's remaining work to do, but this is a good start for people that want to experiment with this on ARC.
Let's get it into v2.0.

**Comment by ioannisg** [comment]
> 
> 
> There's remaining work to do, but this is a good start for people that want to experiment with this on ARC.
> Let's get it into v2.0.

@vonhust have you addressed the latest feedback by @dbkinder ?

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: <the path of normal application elf file> means any zephyr image built in line 73, i will make it more clarified

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: Fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: Fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: Fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: Fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: Fixed

File: samples/boards/arc_secure_services/README.rst
Comment (vonhust) [reply samples/boards/arc_secure_services/README.rst]: Fixed

**Review by dbkinder — DISMISSED** [review]
I see you mention my comments were "fixed", but I don't see the changes actually pushed.

**Comment by vonhust** [comment]
> , but I don't see the changes actually pushed.

I made a force push, could you check that again?

**Comment by ioannisg** [comment]
Seems to me the fixes were pushed (but might be missing something @dkinder).

**Comment by vonhust** [comment]
better to merge after PR #18079 and #18142

**Comment by ioannisg** [comment]
> 
> 
> better to merge after PR #18079 and #18142

@vonhust please resolve the conflicts, after merging the above PRs.
Note that the review needs to conclude today, if this is to get to 2.0 release.

**Comment by vonhust** [comment]
> > better to merge after PR #18079 and #18142
> 
> @vonhust please resolve the conflicts, after merging the above PRs.
> Note that the review needs to conclude today, if this is to get to 2.0 release.

confilct fixed, ok to merge

**Comment by ioannisg** [comment]
> 
> 
> > > better to merge after PR #18079 and #18142
> > 
> > 
> > @vonhust please resolve the conflicts, after merging the above PRs.
> > Note that the review needs to conclude today, if this is to get to 2.0 release.
> 
> confilct fixed, ok to merge

There is still a request-for-changes by @dbkinder.
@vonhust can this PR go into 2.0.0 regardless of what happens with #17767?

**Comment by vonhust** [comment]
> > > > better to merge after PR #18079 and #18142
> > > 
> > > 
> > > @vonhust please resolve the conflicts, after merging the above PRs.
> > > Note that the review needs to conclude today, if this is to get to 2.0 release.
> > 
> > 
> > confilct fixed, ok to merge
> 
> There is still a request-for-changes by @dbkinder.
> @vonhust can this PR go into 2.0.0 regardless of what happens with #17767?

There is no relation between this PR and #17767. I just dismissed dbkinder's request as i addressed his feedbacks. Let's go for merge

**Comment by ioannisg** [comment]
> 
> 
> > > > > better to merge after PR #18079 and #18142
> > > > 
> > > > 
> > > > @vonhust please resolve the conflicts, after merging the above PRs.
> > > > Note that the review needs to conclude today, if this is to get to 2.0 release.
> > > 
> > > 
> > > confilct fixed, ok to merge
> > 
> > 
> > There is still a request-for-changes by @dbkinder.
> > @vonhust can this PR go into 2.0.0 regardless of what happens with #17767?
> 
> There is no relation between this PR and #17767. I just dismissed dbkinder's request as i addressed his feedbacks. Let's go for merge

Yes. @dbkinder if there are still imperfections in documentation, we can definitely address them during the stabilization period.

**Pull request merged**
